### PR TITLE
RemoveRegistryKey

### PR DESCRIPTION
**Pull request by Xaddan**
Hi.
The implementation of the [RemoveRegistryKey](http://wixtoolset.org/documentation/manual/v3/xsd/wix/removeregistrykey.html) element has been added.

### CODE REVIEW

File: Source/src/WixSharp/RemoveRegistryKey.cs [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+		/// <summary>
+		/// Adds itself as an XML content into the WiX source being generated from the <see cref="WixSharp.Project"/>.
+		/// See 'Wix#/samples/Extensions' sample for the details on how to implement this interface correctly.
+		/// </summary>
+		/// <param name="context">The context.</param>
+		public void Process(ProcessingContext context)
+		{
+			context.XParent
+				.FindFirst("Component")
```
Comment (Xaddan) [inline Source/src/WixSharp/RemoveRegistryKey.cs]: It seems not quite right to do so. It turns out that `RemoveRegistryKey` falls into a `Сomponent` that can contain a `File`. In this case `ICE57` does not pass the verification:
```
error LGHT0204: ICE57: Component 'Component.***' has both per-user and per-machine data with a per-machine KeyPath.
```
Here need a separate `Component`.
The same goes for `RemoveRegistryValue`.

Comment (oleg-shilo) [reply Source/src/WixSharp/RemoveRegistryKey.cs]: Are you going to provide correction?
If you do you can now create the component as below:
```C#
XElement component = this.CreateParentComponent();
component.Add(this.ToXElement("RemoveRegistryKey")
                  .AddAttributes($"Root={Root.ToWString()}");
```
And then you can find the appropriate place to insert it.

Comment (Xaddan) [reply Source/src/WixSharp/RemoveRegistryKey.cs]: Unfortunately I do not know how to do it completely. You wrote the easiest, I figured it out myself, but here's how to get the `Directory` into which to add this `Component` (the one that `GetTopLevelDir()` from `Compiler.cs` produces), and also add this `Component` to the `Features`.

Comment (oleg-shilo) [reply Source/src/WixSharp/RemoveRegistryKey.cs]: This part is easy. If you need to find the component "upstream" then you use either of:
- `context.Parent` - WixEntity parent
- `context.Parent` - XElement parent
- `context.Project` - Root level WixEntity

```C#
public void Process(ProcessingContext context)
{
    ...
}
```
You can even use `GetTopLevelDir` if you need to. It's private but I can make it public. As a temp measure you can just implement it itself:
```C#
static XElement GetTopLevelDir(XElement product)
{
    XElement dir = product.Elements("Directory").First();
    XElement prevDir = null;
    while (dir != null)
    {
        prevDir = dir;
        if (dir.Elements("Component").Count() == 0) //just a subdirectory without any installable items
            dir = dir.Elements("Directory").FirstOrDefault();
        else
            return dir; //dir containing installable items (e.g. files or shortcuts)
    }

    return prevDir;
}
```

**Comment by Xaddan** [comment]
What about `Features`?

`context.FeatureComponents.Map(ActualFeatures, Id);`

File: Source/src/WixSharp/RemoveRegistryKey.cs (hunk 1 above)
Comment (Xaddan) [reply Source/src/WixSharp/RemoveRegistryKey.cs]: What about Features?
I did this:
`context.FeatureComponents.Map(ActualFeatures, Id);`
But the component was not added to the list of features.
I looked in `Compiler.cs` and saw that `ProcessFeatures()` goes before processing `IGenericEntity`

I swapped them and earned them, the component was added to the list of features.
```
var context = new ProcessingContext
{
   Project = project,
   Parent = project,
   XParent = product,
   FeatureComponents = featureComponents
};

foreach (IGenericEntity item in project.GenericItems)
   item.Process(context);

ProcessFeatures(project, product, featureComponents, autoGeneratedComponents, defaultFeatureComponents);
```

But this only works if the feature was specified by me, if the feature is not set then the component will not be added to the Feature `Complete`.

Comment (Xaddan) [reply Source/src/WixSharp/RemoveRegistryKey.cs]: Everything, understood.

**Pull request merged**